aea/helpers/async_utils.py

```diff
@@ -0,0 +1,324 @@
+        :param period: period in seconds.
+        :param start_at: optional first call datetime
+        :param exception_callback: optional handler to call on exception raised.
+        :param loop: optional asyncio event loop
+        """
+        self._loop = loop or asyncio.get_event_loop()
+        self._periodic_callable = callback
+        self._start_at = start_at or datetime.datetime.now()
+        self._period = period
+        self._timerhandle: Optional[TimerHandle] = None
+        self._exception_callback = exception_callback
+
+    def _callback(self) -> None:
+        """Call on each scheduled call."""
+        self._schedule_call()
+        try:
+            self._periodic_callable()
+        except Exception as exception:
+            if not self._exception_callback:
+                raise
+            self._exception_callback(self._periodic_callable, exception)
+
+    def _schedule_call(self) -> None:
+        """Set schedule for call."""
+        if self._timerhandle is None:
+            ts = time.mktime(self._start_at.timetuple())
+            delay = max(0, ts - time.time())
+            self._timerhandle = self._loop.call_later(delay, self._callback)
+        else:
+            self._timerhandle = self._loop.call_later(self._period, self._callback)
+
+    def start(self) -> None:
+        """Activate period calls."""
+        if self._timerhandle:
+            return
+        self._schedule_call()
+
+    def stop(self) -> None:
+        """Remove from schedule."""
+        if not self._timerhandle:
+            return
+
+        self._timerhandle.cancel()
+        self._timerhandle = None
+
+
+def ensure_loop(loop: AbstractEventLoop = None) -> AbstractEventLoop:
+    """Use loop provided or create new if not provided or closed."""
+    try:
+        loop = loop or asyncio.get_event_loop()
+        assert not loop.is_closed()
+        assert not loop.is_running()
+    except (RuntimeError, AssertionError):
+        loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(loop)
+    return loop
+
+
+async def wait_and_cancel(
+    tasks: Sequence[Task],
+    include_cancelled: bool = False,
+    loop: Optional[AbstractEventLoop] = None,
+) -> List[Exception]:
+    """
+    Wait first task completed or exception raised and cancel other tasks.
+
+    :param tasks: list of tasks to run and wait.
+
+    :return: list of exceptions raised.
+    """
+    exceptions: List[Exception] = []
+
+    _, pending = await asyncio.wait(tasks, return_when=FIRST_COMPLETED)
+
+    for task in pending:
+        task.cancel()
+
+    completed, pending = await asyncio.wait(tasks, return_when=ALL_COMPLETED)
+
+    assert not pending
+
+    for task in completed:
+        if not include_cancelled and task.cancelled():
+            continue
+        exc = task.exception()
+        if exc:
+            exceptions.append(cast(Exception, exc))
+
+    return exceptions
+
+
+class AnotherThreadTask:
```

Comment:
That is not an appropriate name for a class imo. Can we be more specific?

Reply:
In particular `Another` should be replaced

Reply:
Looking at the implementation, there's no constraint on whether the `loop` provided in `__init__` belongs to another thread or to the same thread (@solarw please correct me if I am mistaken).

If the above is correct, I'd suggest `ThreadSafeTask`, or `ThreadSafeAsyncTask`.

Reply:
Another idea (maybe not today because might require too much extra work): what about inheriting from `asyncio.Future`, or `asyncio.Task`? So we are compliant with those interfaces. We could add a `TODO` for that.